pkg/controller/operators/catalog/deprecated.go

```diff
@@ -0,0 +1,37 @@
+package catalog
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// deprecatedError is returned from installplan execution when a step contains a resource in a GroupVersion that has been deprecated.
+type deprecatedError struct {
+	gv   schema.GroupVersion
+	k    string
+	name string
+}
+
+func (d deprecatedError) Error() string {
+	return fmt.Sprintf("resource has been deprecated: %s of Kind %s is in a deprecated GroupVersion %s and is no longer installable with this version of OLM. %s", d.name, d.k, d.gv,
+		"See https://kubernetes.io/docs/reference/using-api/deprecation-guide/ for information on how to update to a supported version of this resource.")
+}
+
+// deprecatedGroupVersions contains a list of OLM-supported GroupVersions that have been deprecated and are now removed in kube 1.22+.
+// Any resources in these GVs are no longer installable on a 1.22+ cluster. OLM will return an install plan error if a resource in any
+// of these group/versions is found in an installplan step before attempting to create it on-cluster.
+// See https://kubernetes.io/docs/reference/using-api/deprecation-guide/ for notes on deprecated resources.
+// As additional resource groups move from deprecated to removed this list may need to be updated.
```

Comment:
I'm still slightly hesitant around introducing this kind of failure logic upstream as I feel like we should be less restrictive here and produce warning messages and/or alerts that notify there are deprecated resources being used currently in the cluster, vs. fully blocking the creation of these resources downstream. We likely don't have the infrastructure in place and don't want to deal with the code skew, but something to think about going forward as we split up this repository. Last time I checked, we have no explicit support when it comes to OLM versions being mapped to a specific k8s version, so we might be fine and just need to notify of this change in the release notes.

Reply:
These are all deprecated *and removed* apiVersions, right? Catalog operator already generate alerts when a user asks it to install a resource that is only deprecated.

Given that we can know at runtime whether or not these API versions are being served, can this error be generated dynamically? Can a 404 status on a POST/create response indicate anything other than the API is not being served (whether due to deprecation and removal or for some other reason)?

Reply:
> Given that we can know at runtime whether or not these API versions are being served, can this error be generated dynamically? Can a 404 status on a POST/create response indicate anything other than the API is not being served (whether due to deprecation and removal or for some other reason)?

We could, but the impression I got from @kevinrizza was that we can't let the deprecated objects POST to the api-server in the first place, OLM must stop them from being created ahead of time during installplan exeuction. Otherwise the OLM SA will generate alerts on itself for installing removed APIs. 

> These are all deprecated and removed apiVersions, right?
Yes, I was debating whether to call these GroupVersions _deprecated_ or _removed_ in the code. I think deprecated makes sense because from what I can tell that signifies that they were removed, but maybe not. Can change it if it's less confusing.

Reply:
Yeah, that sounds about right last time I had checked. I know there are some e2e checks in the conformance jobs that check (and count) the number of times a serviceaccount accesses a deprecated/remoted API. Those results _should_ report back as a flake so it's easy to miss that in e2e feedback.

It looks like the `[sig-arch][Late] clients should not use APIs that are removed in upcoming releases` reports which serviceaccounts/users/etc. are attempting to access that list of deprecated APIs.

Reply:
I guess the point I was attempting to make was whether it made sense for newer/unreleased versions of upstream OLM to attempt to support a wider range of Kubernetes version, vs. making the inference that certain OLM releases only follow the minor version support path that Kubernetes itself follows. I assume we'd follow the latter given the large support burden of the former, but it would be difficult to release new OLM features to users' running older versions of Kubernetes, e.g. managed Kubernetes platforms lagging behind in minor versions.

Reply:
> We could, but the impression I got from @kevinrizza was that we can't let the deprecated objects POST to the api-server in the first place, OLM must stop them from being created ahead of time during installplan exeuction. Otherwise the OLM SA will generate alerts on itself for installing removed APIs.

There's always the option of querying API discovery for the GVK to avoid sending a request to the endpoint itself. I'd prefer detecting this dynamically b/c:
- we should already be doing something similar for CR instances included in bundles
- I'm really not a fan of maintaining deny lists and managing skew